OBSDOCS-920: Add troubleshooting steps for KubePersistentVolumeFillin… (#73686)
openshift/openshift-docs: modules/monitoring-resolving-the-kubepersistentvolumefillingup-alert-firing-for-prometheus.adoc

// Module included in the following assemblies:
//
// * monitoring/troubleshooting-monitoring-issues.adoc
// * support/troubleshooting/investigating-monitoring-issues.adoc

:_mod-docs-content-type: PROCEDURE
[id="resolving-the-kubepersistentvolumefillingup-alert-firing-for-prometheus_{context}"]
= Resolving the KubePersistentVolumeFillingUp alert firing for Prometheus

As a cluster administrator, you can resolve the `KubePersistentVolumeFillingUp` alert being triggered for Prometheus. 

The critical alert fires when a persistent volume (PV) claimed by a `prometheus-k8s-*` pod in the `openshift-monitoring` project has less than 3% of its total space remaining. This can cause Prometheus to not function normally.

[NOTE]
====

> There are two `KubePersistentVolumeFillingUp` alerts:

Author: @simonpasquier I think it would be better to enhance the [Managing alerts](https://docs.openshift.com/container-platform/4.15/monitoring/managing-alerts.html) section with this information, because I definitely think it is not mentioned explicitly enough (the exact label is mentioned only in one example).

Maybe here it would be better to change it as follows?

* **Critical Alert**: The alert with `critical` severity is triggered when the mounted PV has less than 3% of its total space remaining. 
* **Warning alert**: The alert with `warning` severity  is triggered when the mounted PV has less than 15% of its total space remaining and is expected to fill up within four days.

WDYT?

> There are two `KubePersistentVolumeFillingUp` alerts:

Reviewer: :+1:

* *Critical alert*: It is triggered when the mounted PV has less than 3% of its total space remaining. 
* *Warning alert*: It is triggered when the mounted PV has less than 15% of its total space remaining and is expected to fill up within four days.
====

You can resolve this issue by freeing some space for the PV by removing the three oldest Prometheus time series database (TSDB) blocks.

.Prerequisites

ifndef::openshift-dedicated,openshift-rosa[]
* You have access to the cluster as a user with the `cluster-admin` cluster role.
endif::openshift-dedicated,openshift-rosa[]
ifdef::openshift-dedicated,openshift-rosa[]
* You have access to the cluster as a user with the `dedicated-admin` role.
endif::openshift-dedicated,openshift-rosa[]
* You have installed the OpenShift CLI (`oc`).

.Procedure

> . Remove the three oldest Prometheus TSDB blocks to make more space available for the PV:

Author: Hello @simonpasquier could you please provide more guidance, what do you think is the best solution here? Do you think it would be enough to maybe put a **Warning** note about the possible dangers? Or do you think it is too risky to include this command at all? How should we therefore deal with this user case? Do we possibly have other methods how to deal with the alert issue?

Thank you in advance for your response!

> . Remove the three oldest Prometheus TSDB blocks to make more space available for the PV:

Reviewer: For instance, maybe the users want to copy the data before deleting?

> . Remove the three oldest Prometheus TSDB blocks to make more space available for the PV:

Author: @simonpasquier I see so we can provide the instructions to identify which directories could be removed as you suggested (in that case I would ask if you could provide me with these instructions) 

We can also add a warning with the suggestion to the user to backup their Prometheus data before running the command (do we include any instructions how to do that?)

Let me know what you think, thank you!

> . Remove the three oldest Prometheus TSDB blocks to make more space available for the PV:

Reviewer: Passing this command to the debug pod returns the size of all blocks, sorted from oldest to newest:
 
````
cd /prometheus/;du -hs $(ls -dt */ | grep -Eo "[0-9|A-Z]{26}")
````

Then users should be able to make a decision on which and how many blocks they need to delete.

Re backup, we don't have a strong history currently so maybe we should just omit this part.

> . Remove the three oldest Prometheus TSDB blocks to make more space available for the PV:

Author: Hello @simonpasquier thank you for the command, when I tried to run it, I encountered the following error messages:
```
$ oc debug prometheus-k8s-0 -n openshift-monitoring --image=$(oc get po -n openshift-monitoring 
prometheus-k8s-0 -o jsonpath='{.spec.containers[?(@.name=="prometheus")].image}') -- 
cd /prometheus/;du -hs $(ls -dt */ | grep -Eo "[0-9|A-Z]{26}")

Starting pod/prometheus-k8s-0-debug-bsk5z ...

Removing debug pod ...
du: cannot read directory './.local/share/containers/storage/overlay/
ed1cc9773f4af7028c34ddf6bdd4f2c33b5ad56f5d07c229bb68b3f9f5805cc0': Permission denied
du: cannot read directory './.local/share/containers/storage/overlay/
ef9a0b462fba48388c3920a19b8572958ecc15d781618022e3a9d8f751b5b8ac': Permission denied
du: cannot read directory './.local/share/containers/storage/overlay-containers/
e843b9252161371d561ad012319a6a7e5344fc296a2aa71621c4ff029fe2b7a7/userdata': Permission denied
27G	         .
```
But when I run it like this:
```
$ oc debug prometheus-k8s-0 -n openshift-monitoring --image=$(oc get po -n openshift-monitoring 
prometheus-k8s-0 -o jsonpath='{.spec.containers[?(@.name=="prometheus")].image}')

Starting pod/prometheus-k8s-0-debug-7xj2w ...

sh-5.1$ cd /prometheus/;du -hs $(ls -dt */ | grep -Eo "[0-9|A-Z]{26}")
506M	         .
```
It ran without any of the errors. 

Do you know why the first method did not work as expected?
Thank you!

> . Remove the three oldest Prometheus TSDB blocks to make more space available for the PV:

Author: Oh yes, that was the issue, thank you!

+
[source,terminal]
----
$ oc debug <prometheus_k8s_pod_name> -n openshift-monitoring \// <1>

> -c prometheus --image=$(oc get po <prometheus_k8s_pod_name> \// <1>

Reviewer 2: ```suggestion
-c prometheus --image=$(oc get po <prometheus_k8s_pod_name -n openshift-monitoring> \// <1>
```

> -o jsonpath='{.spec.containers[0].image}') \

Reviewer 2: ```suggestion
| jq '.spec.containers[] |select(.name=="prometheus").image' -r\
```

> -o jsonpath='{.spec.containers[0].image}') \

Reviewer 2: ```suggestion
-o jsonpath='{.spec.containers[?(@.name=="prometheus")].image}') \
```
need to strict to prometheus image, since we can not guarantee the first container (spec.containers[0]) is always prometheus

> -- sh -c 'ls -latr /prometheus/ | egrep -o "[0-9|A-Z]{26}" | head -3 | \

Reviewer 2: is the `egrep -o "[0-9|A-Z]{26}"` right, seems it's wrong, need the dev for confirm

Reviewer 2: no issue now
```
$ oc -n openshift-monitoring rsh -c prometheus prometheus-k8s-0
sh-5.1$ ls /prometheus
01HTYTFRZHJAZ8WPP1XZTTFHB0  chunks_head  lock  queries.active  wal
```
`01HTYTFRZHJAZ8WPP1XZTTFHB0` matches `egrep -o "[0-9|A-Z]{26}"`
https://prometheus.io/docs/prometheus/latest/storage/#on-disk-layout also prove it

* *Warning alert*: The alert with the `severity="warning"` label is triggered when the mounted PV has less than 15% of its total space remaining and is expected to fill up within four days.
====

To address this issue, you can remove Prometheus time series database (TSDB) blocks to create more space for the PV.

.Prerequisites

ifndef::openshift-dedicated,openshift-rosa[]
* You have access to the cluster as a user with the `cluster-admin` cluster role.
endif::openshift-dedicated,openshift-rosa[]
ifdef::openshift-dedicated,openshift-rosa[]
* You have access to the cluster as a user with the `dedicated-admin` role.
endif::openshift-dedicated,openshift-rosa[]
* You have installed the OpenShift CLI (`oc`).

.Procedure

. List the size of all TSDB blocks, sorted from oldest to newest:
+
[source,terminal]
----
$ oc debug <prometheus_k8s_pod_name> -n openshift-monitoring \// <1>
-c prometheus --image=$(oc get po -n openshift-monitoring <prometheus_k8s_pod_name> \// <1>
-o jsonpath='{.spec.containers[?(@.name=="prometheus")].image}') \
-- sh -c 'cd /prometheus/;du -hs $(ls -dt */ | grep -Eo "[0-9|A-Z]{26}")'
----

> <1> Replace `<prometheus_k8s_pod_name>` with the pod mentioned in the `KubePersistentVolumeFillingUp` alert description.

Author: Hello @juzhao could you please help me with the output for this command?
With my simple testing I only got very simple output that looked like this:
```
Starting pod/prometheus-k8s-0-debug-7hjzp ...
72M	.
```
Would you help me provide more complex output that would be close to real example? I think that would be very beneficial to include.
Thank you very much in advance!

Reviewer: ```
308K    01HVKMPKQWZYWS8WVDAYQHNMW6
5.2M    01HVK64DTDA81799TBR9QDECEZ
1.2M    01HVK64DS7TRZRWF2756KHST5X
140K    01HVJS59K11FBVAPVY57K88Z11
9.0M    01HVH2A5Z58SKT810EM6B9AT50
1.2M    01HV8ZDVQMX41MKCN84S32RRZ1
3.4M    01HV6Q2N26BK63G4RYTST71FBF
156K    01HV664H9J9Z1FTZD73RD1563E
16K     01HTHXB60A7F239HN7S2TENPNS
104K    01HTHMGRXGS0WXA3WATRXHR36B
````

Author: Simon, thank you very much for the output! :heart:

> . Identify which and how many blocks could be removed. For example, to remove the three oldest Prometheus TSDB blocks, run the following command:
> +
> [source,terminal]
> ----
> $ oc debug <prometheus_k8s_pod_name> -n openshift-monitoring \// <1>
> -c prometheus --image=$(oc get po -n openshift-monitoring <prometheus_k8s_pod_name> \// <1>
> -o jsonpath='{.spec.containers[?(@.name=="prometheus")].image}') \
> -- sh -c 'ls -latr /prometheus/ | egrep -o "[0-9|A-Z]{26}" | head -3 | \
> while read BLOCK; do rm -rf /prometheus/$BLOCK; done'

Reviewer: (nit) I would avoid the `force/-f` flag.
```suggestion
. Identify which and how many blocks could be removed. For example, to remove the three oldest Prometheus TSDB blocks, run the following command:
+
[source,terminal]
----
$ oc debug <prometheus_k8s_pod_name> -n openshift-monitoring \// <1>
-c prometheus --image=$(oc get po -n openshift-monitoring <prometheus_k8s_pod_name> \// <1>
-o jsonpath='{.spec.containers[?(@.name=="prometheus")].image}') \
-- sh -c 'ls -latr /prometheus/ | egrep -o "[0-9|A-Z]{26}" | head -3 | \
while read BLOCK; do rm -r /prometheus/$BLOCK; done'
```

// Module included in the following assemblies:
//
// * monitoring/troubleshooting-monitoring-issues.adoc
// * support/troubleshooting/investigating-monitoring-issues.adoc

:_mod-docs-content-type: PROCEDURE
[id="resolving-the-kubepersistentvolumefillingup-alert-firing-for-prometheus_{context}"]
= Resolving the KubePersistentVolumeFillingUp alert firing for Prometheus

As a cluster administrator, you can resolve the `KubePersistentVolumeFillingUp` alert being triggered for Prometheus. 

> The critical alert fires when a persistent volume (PV) claimed by a `prometheus-k8s-*` pod in the `openshift-monitoring` project has less than 3% of its total space remaining. This can cause Prometheus to not function normally.

Reviewer 3: ```suggestion
The critical alert fires when a persistent volume (PV) claimed by a `prometheus-k8s-*` pod in the `openshift-monitoring` project has less than 3% total space remaining. This can cause Prometheus to function abnormally.
```

[NOTE]
====
There are two `KubePersistentVolumeFillingUp` alerts:

> * *Critical alert*:  The alert with the `severity="critical"` label is triggered when the mounted PV has less than 3% of its total space remaining.

Reviewer 3: ```suggestion
* *Critical alert*:  The alert with the `severity="critical"` label is triggered when the mounted PV has less than 3% total space remaining.
```

* *Warning alert*: The alert with the `severity="warning"` label is triggered when the mounted PV has less than 15% of its total space remaining and is expected to fill up within four days.
====

> To address this issue, you can remove Prometheus time series database (TSDB) blocks to create more space for the PV.

Reviewer 3: ```suggestion
To address this issue, you can remove Prometheus time-series database (TSDB) blocks to create more space for the PV.
```

The critical alert fires when a persistent volume (PV) claimed by a `prometheus-k8s-*` pod in the `openshift-monitoring` project has less than 3% of its total space remaining. This can cause Prometheus to not function normally.

[NOTE]
====
There are two `KubePersistentVolumeFillingUp` alerts:

* *Critical alert*:  The alert with the `severity="critical"` label is triggered when the mounted PV has less than 3% of its total space remaining.
* *Warning alert*: The alert with the `severity="warning"` label is triggered when the mounted PV has less than 15% of its total space remaining and is expected to fill up within four days.
====

To address this issue, you can remove Prometheus time series database (TSDB) blocks to create more space for the PV.

.Prerequisites

ifndef::openshift-dedicated,openshift-rosa[]
* You have access to the cluster as a user with the `cluster-admin` cluster role.
endif::openshift-dedicated,openshift-rosa[]
ifdef::openshift-dedicated,openshift-rosa[]
* You have access to the cluster as a user with the `dedicated-admin` role.
endif::openshift-dedicated,openshift-rosa[]
* You have installed the OpenShift CLI (`oc`).

.Procedure

> . List the size of all TSDB blocks, sorted from oldest to newest:

Reviewer 4: This step needs some explicit phrasing about running a command, per our [docs guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#procedure-step-that-introduces-a-command):
```suggestion
. List the size of all TSDB blocks, sorted from oldest to newest, by running the following command:
```

+
[source,terminal]
----
$ oc debug <prometheus_k8s_pod_name> -n openshift-monitoring \// <1>
-c prometheus --image=$(oc get po -n openshift-monitoring <prometheus_k8s_pod_name> \// <1>
-o jsonpath='{.spec.containers[?(@.name=="prometheus")].image}') \
-- sh -c 'cd /prometheus/;du -hs $(ls -dt */ | grep -Eo "[0-9|A-Z]{26}")'
----
<1> Replace `<prometheus_k8s_pod_name>` with the pod mentioned in the `KubePersistentVolumeFillingUp` alert description.
+
.Example output
[source,terminal]
----
308M    01HVKMPKQWZYWS8WVDAYQHNMW6
52M     01HVK64DTDA81799TBR9QDECEZ
102M    01HVK64DS7TRZRWF2756KHST5X
140M    01HVJS59K11FBVAPVY57K88Z11
90M     01HVH2A5Z58SKT810EM6B9AT50
152M    01HV8ZDVQMX41MKCN84S32RRZ1
354M    01HV6Q2N26BK63G4RYTST71FBF
156M    01HV664H9J9Z1FTZD73RD1563E

> 216K     01HTHXB60A7F239HN7S2TENPNS

Reviewer 2: ```suggestion
216M    01HTHXB60A7F239HN7S2TENPNS
```

> 104K    01HTHMGRXGS0WXA3WATRXHR36B

Reviewer 2: checked in my cluster, seems the filesize is larger than Simon's example
```
153M	01HVTBKJQV3N59130E5492Z9XT
71M	01HVT9V34H92G7A1G4C7HRZ930
```

> 308K    01HVKMPKQWZYWS8WVDAYQHNMW6
> 5.2M    01HVK64DTDA81799TBR9QDECEZ
> 1.2M    01HVK64DS7TRZRWF2756KHST5X
> 140K    01HVJS59K11FBVAPVY57K88Z11
> 9.0M    01HVH2A5Z58SKT810EM6B9AT50
> 1.2M    01HV8ZDVQMX41MKCN84S32RRZ1
> 3.4M    01HV6Q2N26BK63G4RYTST71FBF
> 156K    01HV664H9J9Z1FTZD73RD1563E
> 16K     01HTHXB60A7F239HN7S2TENPNS
> 104K    01HTHMGRXGS0WXA3WATRXHR36B

Reviewer: more "plausible" values
```suggestion
308M    01HVKMPKQWZYWS8WVDAYQHNMW6
52M     01HVK64DTDA81799TBR9QDECEZ
102M    01HVK64DS7TRZRWF2756KHST5X
140M    01HVJS59K11FBVAPVY57K88Z11
90M     01HVH2A5Z58SKT810EM6B9AT50
152M    01HV8ZDVQMX41MKCN84S32RRZ1
354M    01HV6Q2N26BK63G4RYTST71FBF
156M    01HV664H9J9Z1FTZD73RD1563E
216K     01HTHXB60A7F239HN7S2TENPNS
104M    01HTHMGRXGS0WXA3WATRXHR36B
```

----

> . Identify which and how many blocks could be removed. For example, to remove the three oldest Prometheus TSDB blocks, run the following command:

Reviewer 4: If I'm interpreting this correctly, I think identifying the blocks you can remove and actually removing them are two separate actions. There are three ways you can approach this, up to you which one you'd prefer:

Option 1 - Creating two separate steps:
```suggestion
. Identify which and how many blocks could be removed. 

. Remove the blocks. For example, to remove the three oldest Prometheus TSDB blocks, run the following command:
```

Option 2 - Explicitly mentioning the removal in the step (these two actions are small and related enough where they could probably be on the same step):
```suggestion
. Identify which and how many blocks could be removed, then remove the blocks. For example, to remove the three oldest Prometheus TSDB blocks, run the following command:
```

Option 3 - Treating the identification action as in implicit action (kind of awkward wording though):
```suggestion
. Remove the blocks you want to remove. For example, to remove the three oldest Prometheus TSDB blocks, run the following command:
```

Author: Good point, thank you for noticing, I really like the second suggestion as it keeps it clean while making it more clear and still easy to read, thank you!

+
[source,terminal]
----

> $ oc debug <prometheus_k8s_pod_name> -n openshift-monitoring \// <1>
> -c prometheus --image=$(oc get po -n openshift-monitoring <prometheus_k8s_pod_name> \// <1>
> -o jsonpath='{.spec.containers[?(@.name=="prometheus")].image}') \
> -- sh -c 'ls -latr /prometheus/ | egrep -o "[0-9|A-Z]{26}" | head -3 | \
> while read BLOCK; do rm -r /prometheus/$BLOCK; done'
> ----
> <1> Replace `<prometheus_k8s_pod_name>` with the pod mentioned in the `KubePersistentVolumeFillingUp` alert description.

Reviewer 4: Not super necessary, but since this is only an example, you could replace the input values with an example name if you wanted. Not sure what the correct syntax for a pod name is, though, so it might not be worth the effort if looking that up would take a long time:
```suggestion
$ oc debug prometheus-k8s-example-pod -n openshift-monitoring \
-c prometheus --image=$(oc get po -n openshift-monitoring prometheus-k8s-example-pod \
-o jsonpath='{.spec.containers[?(@.name=="prometheus")].image}') \
-- sh -c 'ls -latr /prometheus/ | egrep -o "[0-9|A-Z]{26}" | head -3 | \
while read BLOCK; do rm -r /prometheus/$BLOCK; done'
----
```

I'll leave that up to you.

Author: Yes, you are right, maybe giving the complete example would be better here :+1:

> . Verify the usage of the mounted PV and make sure there is enough space available:

Reviewer 4: Same comment as above about mentioning a command:
```suggestion
. Verify the usage of the mounted PV and make sure there is enough space available by running the following command:
```
